File: commonfamily/src/main/java/com/koenig/commonModel/finance/CostDistribution.java
package com.koenig.commonModel.finance;

import com.koenig.commonModel.Byteable;
import com.koenig.commonModel.User;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by Thomas on 22.10.2017.
 */

public class CostDistribution extends Byteable {
    Map<User, Costs> distribution;

    public CostDistribution(Map<User, Costs> distribution) {
        this.distribution = distribution;
    }

    public CostDistribution() {
        distribution = new HashMap<>();
    }

    public CostDistribution(ByteBuffer buffer) {
        this();
        int size = buffer.getInt();
        for (int i = 0; i < size; i++) {
            User user = new User(buffer);
            Costs costs = new Costs(buffer);
            distribution.put(user, costs);
        }
    }

    public Map<User, Costs> getDistribution() {
        return distribution;
    }

    public int sumReal() {
        int sumReal = 0;
        for (Costs c : distribution.values()) {
            sumReal += c.Real;
        }

        return sumReal;
    }

    public int sumTheory() {
        int sumTheory = 0;
        for (Costs c : distribution.values()) {
            sumTheory += c.Theory;
        }

        return sumTheory;
    }

    public boolean isValid() {
        return sumReal() == sumTheory();
    }

    public Costs getCostsFor(User user) {
        Costs costs = distribution.get(user);
        return costs == null ? new Costs(0, 0) : costs;
    }


    public void putCosts(User user, Costs costs) {
        distribution.put(user, costs);
    }

    public void putCosts(User user, int real, int theory) {
        putCosts(user, new Costs(real, theory));
    }

    @Override
    public String toString() {
        if (distribution.isEmpty()) {
            return "Empty CostDistribution";
        }
        StringBuilder builder = new StringBuilder();
        for (User user : distribution.keySet()) {
            Costs costs = distribution.get(user);
            builder.append(user.getName());
            builder.append(": ");
            builder.append(costs);
            builder.append("\n");
        }

        return builder.substring(0, builder.length() - 1);
    }

    @Override
    public int getByteLength() {
        int size = 4;
        for (User user : distribution.keySet()) {
            size += user.getByteLength() + distribution.get(user).getByteLength();
        }

        return size;
    }

    public float getRealPercent(User user) {
        int sum = sumReal();
        if (sum == 0) {
            return 0;
        }

        return getCostsFor(user).Real / (float) sum;
    }

    public float getTheoryPercent(User user) {
        int sum = sumTheory();
        if (sum == 0) {
            return 0;
        }

        return getCostsFor(user).Theory / (float) sum;
    }

    @Override
    public void writeBytes(ByteBuffer buffer) {
        buffer.putInt(distribution.size());
        for (User user : distribution.keySet()) {
            user.writeBytes(buffer);
            buffer.put(distribution.get(user).getBytes());
        }
    }
}

File: commonfamily/src/main/java/com/koenig/commonModel/finance/CostDistribution.kt
package com.koenig.commonModel.finance

import com.koenig.commonModel.Byteable
import com.koenig.commonModel.User
import java.nio.ByteBuffer
import java.util.*

/**
 * Created by Thomas on 22.10.2017.
 */

class CostDistribution : Byteable {
    internal var distribution: MutableMap<User, Costs>

    val isValid: Boolean
        get() = sumReal() == sumTheory()

    override val byteLength: Int
        get() {
            var size = 4
            for (user in distribution.keys) {
                size += user.byteLength + distribution[user]!!.byteLength
            }

            return size
        }

    constructor(distribution: MutableMap<User, Costs>) {
        this.distribution = distribution
    }

    constructor() {
        distribution = HashMap()
    }

    constructor(buffer: ByteBuffer) : this() {
        val size = buffer.int
        for (i in 0 until size) {
            val user = User(buffer)
            val costs = Costs(buffer)
            distribution[user] = costs
        }
    }

    fun getDistribution(): Map<User, Costs> {
        return distribution
    }

    fun sumReal(): Int {
        var sumReal = 0
        for (c in distribution.values) {
            sumReal += c.Real
        }

        return sumReal
    }

    fun sumTheory(): Int {
        var sumTheory = 0
        for (c in distribution.values) {
            sumTheory += c.Theory
        }

        return sumTheory
    }

    fun getCostsFor(user: User): Costs {
        val costs = distribution[user]
        return costs ?: Costs(0, 0)
    }


    fun putCosts(user: User, costs: Costs) {
        distribution[user] = costs
    }

    fun putCosts(user: User, real: Int, theory: Int = real) {
        putCosts(user, Costs(real, theory))
    }

    override fun toString(): String {
        if (distribution.isEmpty()) {
            return "Empty CostDistribution"
        }
        val builder = StringBuilder()
        for (user in distribution.keys) {
            val costs = distribution[user]
            builder.append(user.name)
            builder.append(": ")
            builder.append(costs)
            builder.append("\n")
        }

        return builder.substring(0, builder.length - 1)
    }

    fun getRealPercent(user: User): Float {
        val sum = sumReal()
        return if (sum == 0) {
            0f
        } else getCostsFor(user).Real / sum.toFloat()

    }

    fun getTheoryPercent(user: User): Float {
        val sum = sumTheory()
        return if (sum == 0) {
            0f
        } else getCostsFor(user).Theory / sum.toFloat()

    }

    override fun writeBytes(buffer: ByteBuffer) {
        buffer.putInt(distribution.size)
        for (user in distribution.keys) {
            user.writeBytes(buffer)
            buffer.put(distribution[user]!!.bytes)
        }
    }

    operator fun get(user: User): Costs {
        return getCostsFor(user)
    }

    operator fun set(user: User, costs: Costs) {
        distribution[user] = costs
    }
}
